### DISCUSSION TITLE
Design Lockup Linear to support Cliff Linear vesting

### DISCUSSION THREAD

**Discussion by smol-ninja** [sablier-labs/evm-monorepo#903] [Ideas]
> [!IMPORTANT]  
> The below proposal is for future release of Lockup and not v2.2.

Context
---------
In the current design of Lockup linear, the amount unlocked at `cliff` is a function of the time elapsed. That is, if the cliff timestamp is 30% ahead of the start time, this would unlock 30% of the total amount at the cliff. This makes it less flexible for creating a cliff linear vesting.

> [!NOTE]  
> In a cliff linear vesting, token distributions include a preset amount unlocked at the cliff followed by a linear distribution. The amount unlocks at the cliff can be any and does not need to follow the curve slope.

Existing solution
---------
A Cliff linear vesting can be created using Lockup Dynamic (LD). It has a few problems:
1. It is not intuitive to use LD for creating Linear streams with cliffs.
2. Creating a cliff linear vesting stream is less gas efficient (same argument as [why we chose to implement lockup tranched](https://github.com/sablier-labs/company-discussions/discussions/19)). 
3. Despite having a cliff feature in Lockup Linear (LL), it does not support Cliff linear vesting. 

Proposed solution
---------
Modify the Lockup Linear design to allow the creation of cliff linear vesting streams. So that,
1. LL can be used to create streams with unlocks instantly followed by a linear curve (as requested by one of our users). 
2. LL can be used to create a stream with a fixed amount of unlock at a cliff time followed by a linear curve.
3. it becomes more gas-efficient than doing the same with LD.
4. Existing curves such as Unlock-cliff and Unlock-linear on Sablier UI will become gas efficient. Because they are created using LD at present. 

Thoughts @sablier-labs/everybody?

**Comment by smol-ninja** [comment]
Supporting examples from [SAFE distribution](https://safe.global/blog/safe-tokenomics):

- 25M tokens are available immediately, with the remaining 25M vested over 4 years
- An initial airdrop of 25M tokens, with half available immediately and the rest vested over 4 years

**Comment by PaulRBerg** [comment]
Sorry for taking so long to respond here!

This is a great idea. Before reading your post, I thought that a proper cliff has to be obey the mathematical rules of the rest of the curve, i.e., in the case of LL to be a function of elapsed time as it is currently implemented. However, [ChatGPT proved me wrong](https://chatgpt.com/share/bd86e176-a188-4adf-87bd-47940ff3f1cc).

Am I right that your proposal has two parts:

1. Free the cliff in Lockup Linear from having to be a function of elapsed time
2. Add an optional amount that gets instantly unlocked (related: https://github.com/sablier-labs/v2-core/discussions/912)

If yes, I am in favor of the proposal. It makes a lot of sense.

Tagging @sablier-labs/engineers for additional feedback. I think we should include this change in Lockup v1.3.0

**Reply by smol-ninja** [reply]
Thanks for taking time to review the proposal. To answer your questions:

> Free the cliff in Lockup Linear from having to be a function of elapsed time

Yes.

> Add an optional amount that gets instantly unlocked

Yes. Instead of amount, we might also add two percentages representing: (i) instant unlocked amount (ii) cliff amount.

```python
if block.timestamp < cliffTime:
    streamedAmount = deposited * instantUnlockPct # instant unlocked amount
else:
    # elapsed time is defined from cliff time
    elapsedTime = (block.timestamp - cliffTime) / (endTime - cliffTime)
    streamedAmount = deposited * instantUnlockPct # instant unlocked amount
                   + deposited * cliffUnlockPct  # cliff unlocked amount
                    # stream linearly from cliff time
                   + deposited * (1 - instantUnlockPct - cliffUnlockPct) * elapsedTime
```

**Reply by PaulRBerg** [reply]
Got it, thanks for confirming.

While these are cool ideas and I agree that implementing them would be beneficial, I also realize now that the `LockupLinear` name would be misleading. With the new design, at least a portion of the curve would be non-linear.

I can't think of any neat name off the top of my head. Maybe `LockupBasic`?

P.S. this would also have implications for the [URL aliases in the UI](https://docs.sablier.com/api/subgraphs/protocol/structure#aliases), i.e. `LL`, `LL2`, etc.

cc @sablier-labs/frontend

**Reply by smol-ninja** [reply]
Even though it's piecewise linear, I don't think `LockupLinear` would be misleading. The terms `Linear`, `Dynamic`, and `Tranched` can represent (or we can make them represent) the continuous streaming portion of the stream, where cliff and instant unlocks are a part of it. If there is no cliff, then they are fully linear and in the presence of a cliff, they are piecewise linear.

I'm looking forward to hearing others' opinions on this.

**Reply by PaulRBerg** [reply]
Makes sense. Happy to keep `LockupLinear` in this case.

There's also the backward-compatibility cost (renaming LL would be a pain for the subgraph, etc.)

**Reply by razgraf** [reply]
I agree with @smol-ninja. While `Linear` is the core part, cliffs and initial unlocks can be defined as "features".

**Comment by razgraf** [comment]
Your suggestion sounds great! Fully support this for our next EVM Lockup release.

**Comment by PaulRBerg** [comment]
Created an issue to track this:

https://github.com/sablier-labs/v2-core/issues/1043

**Comment by smol-ninja** [comment]
A user requested unlock-cliff with 0 cliff amount. Post this implementation, we will also be able to offer such kind of streams through airstream campaigns which is not possible today due to lack of LD airstreams cc @maxdesalle.

<img width="468" alt="Screenshot 2024-10-30 at 20 26 33" src="https://github.com/user-attachments/assets/a3fe109c-6fab-4197-b9e0-6b433a25d2f8">

**Reply by PaulRBerg** [reply]
Great stuff.

However, in order to provide a general-purpose solution for requests like this, we should implement a fully customizable [segment generator UI](https://github.com/sablier-labs/v2-interfaces/discussions/577).
